include/interfaces/ALE_navier_stokes.h

```diff
@@ -100,8 +217,135 @@ template <int dim, int spacedim, typename LAC>
 void ALENavierStokes<dim,spacedim,LAC>::
 set_matrix_couplings(std::vector<std::string> &couplings) const
 {
-  couplings[0] = "1,1,1; 1,1,1; 1,1,1"; // TODO: Select only not null entries
-  couplings[1] = "0,0,0; 0,0,0; 0,0,1";
+  couplings[0] = "1,1,1,1; 1,1,1,1; 1,1,1,1; 1,1,1,1"; // TODO: Select only not null entries
+  couplings[1] = "0,0,0,0; 0,0,0,0; 0,0,0,0; 0,0,0,1";
+}
+
+template <int dim, int spacedim, typename LAC>
+void
+ALENavierStokes<dim,spacedim,LAC>::
+solution_preprocessing(FEValuesCache<dim,spacedim> &fe_cache) const
+{
+  Tensor<1, dim, double> global_force;
+
+  typedef
+  FilteredIterator<typename DoFHandler<dim, spacedim>::active_cell_iterator>
+  CellFilter;
+
+  auto local_copy = [this, &global_force]
+                    (const CopyForce<dim> &data)
+  {
+    global_force += data.local_force;
+  };
+
+  auto local_assemble = [ this, &fe_cache]
+                        (const typename DoFHandler<dim, spacedim>::active_cell_iterator & cell,
+                         FEValuesCache<dim,spacedim> &scratch,
+                         CopyForce<dim> &data)
+  {
+    assemble_local_forces(cell, scratch, data, fe_cache);
+  };
+
+  WorkStream::
+  run (CellFilter (IteratorFilters::LocallyOwnedCell(),
+                   this->dof_handler->begin_active()),
+       CellFilter (IteratorFilters::LocallyOwnedCell(),
+                   this->dof_handler->end()),
+       local_assemble,
+       local_copy,
+       fe_cache,
+       CopyForce<dim>());
+
+  auto &cache = fe_cache.get_cache();
```

Comment:
Su questo ho fatto un po’ di tentativi. Quella che non mi da errori (di compilazione) è la versione che ho implementato.
L’errore di seg fault mi viene se calcolo una forza che dipenda da u e p…


On 16 February 2016 at 22:40:06, Alberto Sartori (notifications@github.com) wrote:

In include/interfaces/ALE_navier_stokes.h:

> +                         CopyForce<dim> &data)
> +  {
> +    assemble_local_forces(cell, scratch, data, fe_cache);
> +  };
> +
> +  WorkStream::
> +  run (CellFilter (IteratorFilters::LocallyOwnedCell(),
> +                   this->dof_handler->begin_active()),
> +       CellFilter (IteratorFilters::LocallyOwnedCell(),
> +                   this->dof_handler->end()),
> +       local_assemble,
> +       local_copy,
> +       fe_cache,
> +       CopyForce<dim>());
> +
> +  auto &cache = fe_cache.get_cache();
solo un pensiero. in deal2lkit bisogna fare

return cache;

oppure

return &cache;
```?

—
Reply to this email directly or view it on GitHub.